### ISSUE TITLE
Can't assign port fp-443 to Private IP Listener already assigned to Public IP Listener

### ISSUE THREAD

**Issue by mhsh64** [Azure/application-gateway-kubernetes-ingress#948]
**Describe the bug**

I have Ingress resources on a AKS cluster which needs using PrivateIp or PublicIp as their front end IP. I have configured one Ingress using public IP, which was fine, however, when using private IP annotation (appgw.ingress.kubernetes.io/use-private-ip: "true") in another Ingress resource, I am getting the bellow error on AGIC:
 
 `E0715 15:13:23.665442       1 frontend_listeners.go:46] Can't assign port fp-443 to Private IP Listener fl-d84d98ab259764775ce8700847d97b66; already assigned to Public IP Listener fl-4db314dd864524bfe3bbc9786686cce0; Will not create listener {FrontendPort:443 HostNames:[ myapp-private.mydomain.com   ] UsePrivateIP:true}
E0715 15:13:23.665476       1 frontend_listeners.go:46] Can't assign port port_80 to Private IP Listener fl-7ad98dd3ce13c2997644bffc6708100d; already assigned to Public IP Listener fl-658b918ffa5ca2a4cab5e0d72e383d46; Will not create listener {FrontendPort:80 HostNames:[ myapp-private.mydomain.com  ] UsePrivateIP:true} `

 And it did not assign any Listeners on Private IP.

 

Please advise is this possible using Public and Private frontend IP at the same time for same ports on one AppGW?

Ingress1
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: aspnetapp
  namespace: ns1
  annotations:
    kubernetes.io/ingress.class: azure/application-gateway
    appgw.ingress.kubernetes.io/appgw-ssl-certificate: rbccert
    appgw.ingress.kubernetes.io/ssl-redirect: "true"
    appgw.ingress.kubernetes.io/use-private-ip: "true"
spec:
  rules:
  - host: myapp-private.mydomain.com
    http:
      paths:
      - path: /
        backend:
          serviceName: aspnetapp
          servicePort: 80

Ingress2

apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: kyvos-network-utility
  namespace: kyvos-network-utility
  annotations:
    kubernetes.io/ingress.class: azure/application-gateway
    appgw.ingress.kubernetes.io/appgw-ssl-certificate: rbccert
    appgw.ingress.kubernetes.io/ssl-redirect: "true"
spec:
  rules:
  - host: myapp.mydomain.com
    http:
      paths:
      - backend:
          serviceName: myapp
          servicePort: 80
        path: /

AGIC version: 1.2.0-rc2
AppGW: WAF V2

**To Reproduce**

**Ingress Controller details**
* Output of `kubectl describe pod <ingress controller`> . The <ingress controller> pod name can be obtained by running `helm list`. 
* Output of `kubectl logs <ingress controller>. 
* Any Azure support tickets associated with this issue.

**Comment by 3quanfeng** [comment]
Hi @mhsh64 , correct, right now AGIC doesn't support[ multiple IPs on the same port](https://github.com/Azure/application-gateway-kubernetes-ingress/blob/master/docs/annotations.md#use-private-ip).

**Comment by mhsh64** [comment]
@3quanfeng 
Thanks
So at the moment if we want to configure this, we will need two AppGw and two AGIC on the cluster with enabling watch separate namespaces?

**Comment by mhsh64** [comment]
@3quanfeng 
Sorry again, it will be very hard and maybe not beneficial to deploy two appGw ( and two AGIC) , one for internal and one for public ingress.
However, if this is the only solution for now, is there any document you can please refer me to deploy two AGIC on a same AKS cluster?
I wish we were able to have one appGw for using both public and private frontend IPs.

Thanks

**Comment by 3quanfeng** [comment]
Hi @mhsh64 , AGIC support for the feature is definitely on our roadmap, but we don't have ETA for it given appgw itself doesn't support it now.
To deploy two AGICs within one AKS cluster but watching on different namespaces, you will need two appgws, and to map one AGIC to one AppGw, you will need to configure in the AGIC helm values.yaml to have the right match. make sure you have 

kubernetes:
  # Namespace(s) AGIC watches; Leaving this blank watches all namespaces;
  # Accepts one or many comma-separated values
  watchNamespace:

# appgw:
#   environment: AZUREPUBLICCLOUD
#   subscriptionId: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
#   resourceGroup: myResourceGroup
#   name: myApplicationGateway
#   # Whether to force private IP for all the listeners on Application Gateway
#   usePrivateIP: false

**Comment by mhsh64** [comment]
@3quanfeng 
Thanks
I deployed two appGws with two AGIC, however it will be really hard to manage, since for each new namespace we have to add them in watched list in AGIC cm.
Isn’t there anyway to make this more usable?

**Comment by 3quanfeng** [comment]
Hi @mhsh64 ,
In the next release version(including rc), AGIC will support to delegate two ingress resources in the same namespace to two different controllers.

**Comment by mhsh64** [comment]
@3quanfeng Thank you for the update
I think it might be also useful rather than using watchNamespace in AGIC configmap, using annotations for Ingress resources to let them know which Ingress Controller they can use. So it will be much easier and manageable to have multiple AGICs, since it is not possible yet to have same ports for public and private frontend IPs.

**Comment by 3quanfeng** [comment]
Thanks @mhsh64 , that's exactly the feature we will release soon to make the ingress class configurable.

**Comment by mhsh64** [comment]
@3quanfeng Thank you
Looking forward to it ☺️

**Comment by sbkg0002** [comment]
This would be great!

**Comment by catay** [comment]
Are there some timelines available when we can expect this feature ?  Thanks.

**Comment by schlickspringer** [comment]
@catay the merge request has been already included to support his. however I'm also waiting the release that includes this merge...

**Comment by qinshuang1998** [comment]
> Hi @mhsh64 , AGIC support for the feature is definitely on our roadmap, but we don't have ETA for it given appgw itself doesn't support it now.
> To deploy two AGICs within one AKS cluster but watching on different namespaces, you will need two appgws, and to map one AGIC to one AppGw, you will need to configure in the AGIC helm values.yaml to have the right match. make sure you have
> 
> `kubernetes:
> watchNamespace:
> 
> appgw:
> environment: AZUREPUBLICCLOUD
> subscriptionId: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
> resourceGroup: myResourceGroup
> name: myApplicationGateway
> usePrivateIP: false`

@3quanfeng I'd like to ask you about the Azure support for creating two AGIC in the same namespace now? 

I want to deploy two AGICs within one AKS cluster but watching on same namespaces, not the different namespaces, to resolve this problem of "Can't assign port fp-443 to Private IP Listener already assigned to Public IP Listener". 

Looking forward to your answer, thanks!

**Comment by aelmanaa** [comment]
@3quanfeng do you know when the merge will be released (cfr. @schlickspringer's message)

**Comment by ezYakaEagle442** [comment]
@3quanfeng @mscatyao 
the [AppGW doc](https://docs.microsoft.com/en-us/azure/application-gateway/application-gateway-faq#can-i-use-the-same-port-for-both-public-facing-and-private-facing-listeners) states :

_Can I use the same port for both public-facing and private-facing listeners?
No._

The [AGIC doc](https://docs.microsoft.com/en-us/azure/application-gateway/ingress-controller-private-ip) explains how to use a Private IP, but not how to manage the “public+private” use case and to how to configures different ports for public+private IP  , could you please add in the docs if this scenario is supported or not ?

**Comment by 3quanfeng** [comment]
Hi @ezYakaEagle442 , there is a [note ](https://github.com/Azure/application-gateway-kubernetes-ingress/blob/master/docs/annotations.md#use-private-ip) indicating that "App Gateway doesn't support multiple IPs on the same port"

**Comment by ezYakaEagle442** [comment]
@3quanfeng thanks foryou reply, but this does not answer the question :  how to manage the “public+private” use case with AGIC and to how to configures different ports for public+private IP with AGIC.
This is not explained in the docs and the [ARM template](https://raw.githubusercontent.com/Azure/application-gateway-kubernetes-ingress/master/deploy/azuredeploy.json) does not manage that neither

**Comment by sbkg0002** [comment]
Is there any place where we can track the status of this roadmap item?

**Comment by Phrow** [comment]
Hi, I'm also interested in either the general availability of "multiple ips on same port in app gateway" or anything that makes configuring two app gateways as a workaround easier.
Does anyone know when any of this is going to be available? @3quanfeng maybe?
Thank you in advance!

**Comment by picardsrcd** [comment]
I am also very keen on a resolution or instructions on a workaround. I would like to avoid moving to another ingress controller.

**Comment by schlickspringer** [comment]
@picardsrcd, @Phrow you can check the latest release. I was able to get it working with public/private IP (different ports).

**Comment by picardsrcd** [comment]
> @picardsrcd, @Phrow you can check the latest release. I was able to get it working with public/private IP (different ports).
Do you have a sample configuration? I am using the AGIC AKS addon on a AKS 1.19.7 cluster.


Private Example:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: aspnetapp-priv
  labels:
    app: aspnetapp-priv
spec:
  containers:
  - image: "mcr.microsoft.com/dotnet/core/samples:aspnetapp"
    name: aspnetapp-image
    ports:
    - containerPort: 80
      protocol: TCP

---

apiVersion: v1
kind: Service
metadata:
  name: aspnetapp-priv
  labels:
    app: aspnetapp-priv
  # annotations:
  #   service.beta.kubernetes.io/azure-load-balancer-internal: "true"
spec:
  selector:
    app: aspnetapp-priv
  # type: LoadBalancer
  ports:
  - protocol: TCP
    port: 8080
    targetPort: 80

---

apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: aspnetapp-priv
  annotations:
    kubernetes.io/ingress.class: azure/application-gateway
    # appgw.ingress.kubernetes.io/backend-protocol: http
    appgw.ingress.kubernetes.io/use-private-ip: "true"
    appgw.ingress.kubernetes.io/backend-path-prefix: "/"
    appgw.ingress.kubernetes.io/override-frontend-port: "8080"
spec:
  rules:
  - http:
      paths:
      - path: "/private/asp/*"
        backend:
          serviceName: aspnetapp-priv
          servicePort: 8080

```

Public Example:
```
apiVersion: v1
kind: Pod
metadata:
  name: aspnetapp-auth
  labels:
    app: aspnetapp-auth
spec:
  containers:
  - image: "mcr.microsoft.com/dotnet/core/samples:aspnetapp"
    name: aspnetapp-image
    ports:
    - containerPort: 80
      protocol: TCP

---

apiVersion: v1
kind: Service
metadata:
  name: aspnetapp-auth
spec:
  selector:
    app: aspnetapp-auth
  ports:
  - protocol: TCP
    port: 80
    targetPort: 80

---

apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: aspnetapp-auth
  annotations:
    kubernetes.io/ingress.class: azure/application-gateway
    appgw.ingress.kubernetes.io/backend-protocol: http
    appgw.ingress.kubernetes.io/backend-path-prefix: "/"
spec:
  rules:
  - http:
      paths:
      - path: /auth/* # The asterix is required to have rewriting work
        backend:
          serviceName: aspnetapp-auth
          servicePort: 80

```

**Comment by Phrow** [comment]
> 
> 
> @picardsrcd, @Phrow you can check the latest release. I was able to get it working with public/private IP (different ports).

But this whole thread is about using the **same port** on public and private IP at the same time, no? Different ports should've always worked.

**Comment by angelbulas** [comment]
Is there no support for public and private IP using the same port at the same time yet?

**Comment by mscatyao** [comment]
We're actively working on allowing the same port to be used with the public and private IP. We'll update the thread as we continue to make progress on this capability.

**Comment by TrotmanRalph** [comment]
Any update on this functionality?

**Comment by guerillamos** [comment]
Bump - Curious if there are any updates on this?

**Comment by matt961** [comment]
This limitation is quite strange, seems there should be no technical reason for it not to work, but thanks for looking into it @mscatyao

**Comment by mscatyao** [comment]
This is a limitation on AppGW that we're still working on. I've added this feature request to the roadmap as "In Progress".

**Comment by hieumoscow** [comment]
Do you ETA on preview @mscatyao?

**Comment by ashwinpagarkhed** [comment]
Thanks @JackStromberg. Just wanted to add a +1 to this feature request.

**Comment by kamekazemaster** [comment]
Is there any progress on this one other than the item beeing on the roadmap (it took ~15 months after reporting this issue to land there). Now ~7 months later the only information is that we get future updates anytime this year. Will it be done this year? Will we be able to use AGIC for private and public listeners on the same port? It is just a huge amount of work to work around this problem and deploy a second ingress (and then if it is solved at some point migrate everything). Any kind of ETA would be incredibly helpful in our planning of the production cluster.

Don't understand me wrong. I am very thankful for your hard work and I am sure that this issue is not the only thing you are working on. I am just asking for a bit more transparency if possible. Thank you.

**Comment by JackStromberg** [comment]
Hey @kamekazemaster, appreciate you reaching out.

We hear the community loud and agree with the value on bringing this feature to light. I know it has been a long journey and while I cannot bring an exact ETA yet, there has been movement on this internally.  I won't have an update for another 1-2months, but we will provide more details on the status of this thereafter.

I apologize I cannot provide more details at this time, but we really do appreciate the patience and are looking forward to making the feature available to all customers.

Thank you,
Jack

**Comment by n3wt0n** [comment]
+1 for having the possibility to use the same port on private and public listeners.
We are in the middle of a huge deployment and we stumbled on this.
For the time being we will use AppGateway only for public endpoints, but this limitation is quite strong for us and we'd really need to solve this quickly... otherwise we'll have to fall back to competitor services

**Comment by guyta-essence** [comment]
i am in the process of migrating to AGIC, this is a showstopper any updates or time estimates.

**Comment by guyta-essence** [comment]
> i am in the process of migrating to AGIC, this is a real blocker.

update i installed nginx ingress controller using it for the internal traffic, continuing testing agic for external traffic.

**Comment by montyplattner** [comment]
@JackStromberg, have there been any developments on this since May?

**Comment by JackStromberg** [comment]
Hey Team,

Apologies for the delay an a much overdue update.  We have made some more progress internally, but unfortunately don't have a release announcement to make quite yet.  I can absolutely confirm this is on our radar and are actively working to unblock this.  I will update this once we have more news to share -- please continue to hold us accountable and feel free to keep on "@ing" me.

Thank you for your patience,
Jack

**Comment by emibcn** [comment]
@JackStromberg, have there been any developments on this since Sep?

**Comment by JackStromberg** [comment]
Hey @emibcn,

Yes, we are invested in bringing this feature forward and have made progress internally.  We will be able to share a more detailed update near end of next month.

Appreciate the patience; we understand the importance of getting this unblocked.

Happy new year!
Jack

**Comment by montyplattner** [comment]
@JackStromberg, Have there been any developments on the availability of this feature?

**Comment by reza8iucs** [comment]
I think the product team just enabled this feature as I was able to enable port 443 on both private and public IPs today.... really a great step forward product team!!! way to go!

**Comment by kengibous** [comment]
@reza8iucs in which region?  I am unable to do so in Central US.

**Comment by JackStromberg** [comment]
Hey Folks,

We will provide an official update in the next two weeks to explain the behavior you are seeing :)

Appreciate your patience on this.

Jack

**Comment by reza8i** [comment]
@kengibous  I am in Canada Central and they must have enabled this in the past few days because I tried it last week and it didn't work.Right now, I'm successfully listening to traffic on port 443 on my private and public IP. All tested and works like a chime!

**Comment by kengibous** [comment]
@reza8i - thanks for providing that, still no luck in Central US.  @JackStromberg is this not available in all regions as of yet?  Looking forward to the official update.

**Comment by JackStromberg** [comment]
This feature is not officially released yet. Please do not assume it should be available in all regions.  We will provide another update this month.

Jack

**Comment by antoineozenne** [comment]
@JackStromberg, do you have an update?

**Comment by JackStromberg** [comment]
> @JackStromberg, do you have an update?

Previous statement still stands. We will provide an update later this month.

Jack

**Comment by JackStromberg** [comment]
Hey folks,

We appreciate your patience on this.  As you can tell, this feature is in the process of its release, targeting completion early March.  There are some changes that are still rolling out, so we do recommend holding off from attempting configuration prior to official release.  We will post a service announcement once this functionality is ready here: https://azure.microsoft.com/updates/

I will update this issue with a link to the announcement once available and we can close this one out.

Thank you,
Jack

**Comment by beetel** [comment]
any updates on this?

**Comment by JackStromberg** [comment]
Hey Team,

We are introducing an additional improvement related to the feature configuration and so the broader Azure Updates communication has been delayed this month. The documentation for the feature is out and you can begin to use this directly on Application Gateway via Portal, CLI, PS, Terraform, etc today. Please take note of the NSG rule requirements (if the subnet uses NSG) as creating same port for public and private listeners changes the incoming flow for all the listeners.
https://learn.microsoft.com/azure/application-gateway/configuration-listeners#frontend-port

Feel free to try this out, looking forward to hearing how it goes!

Thank you,
Jack

**Comment by gpltaylor** [comment]
@JackStromberg I am still getting the same issue as others in this post. However, this is erratic. 

I have had Public and Private HTTPS working during development.
However, this morning all Private services have failed when I tried to deploy a new API Service.

```
frontend_listeners.go:46] Can't assign port fp-443 to Private IP Listener fl-19f2f2235084a56da17a1948bc9c0801; already assigned to Public IP Listener fl-4894b96811cb1cfa9b7a55e34cc4e3ce; 
Will not create listener {FrontendPort:443 HostNames:[fmg-tester.fluent-uat-services.private    ] UsePrivateIP:true}
```

Do I need to update the Cluster/Ingress?

```
managed-by-k8s-ingress": "1.5.3/94b2b229/2023-02-03-21:42T+0000
Aks Version: 1.24.6
```

**Comment by JackStromberg** [comment]
Hey Team,

Yes, there is a check in AGIC today on the same port.  I apologize I was not clear; the changes have been made available for Application Gateway and can begin to be used as-is today.  In parallel, we have been working on publishing v1.7 stable for both helm and add-on (helm already available); the changes for same port public/private listener in AGIC will be made available in the next release.

Thank you for your patience,
Jack

**Comment by JackStromberg** [comment]
Hey folks!

Appreciated all the patience on this.  We are pleased to announce this is now available in AGIC v1.7.1, which can be deployed via helm today.  https://github.com/Azure/application-gateway-kubernetes-ingress/releases/tag/1.7.1

Cheers!
Jack
